src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+	RATBAG_BUTTON_TYPE_LEFT,
+	RATBAG_BUTTON_TYPE_RIGHT,
+	RATBAG_BUTTON_TYPE_MIDDLE,
+	RATBAG_BUTTON_TYPE_THUMB,
+	RATBAG_BUTTON_TYPE_THUMB2,
+	RATBAG_BUTTON_TYPE_RESOLUTION_CYCLE_UP,
+	RATBAG_BUTTON_TYPE_THUMB3,
+	RATBAG_BUTTON_TYPE_THUMB4,
+	RATBAG_BUTTON_TYPE_WHEEL_UP,
+	RATBAG_BUTTON_TYPE_WHEEL_DOWN,
+};
+
+struct gskill_button_function_mapping {
+	enum gskill_button_function_type type;
+	struct ratbag_button_action action;
+};
+
+static const struct gskill_button_function_mapping gskill_button_function_mapping[] = {
+	{ GSKILL_BUTTON_FUNCTION_MACRO,              BUTTON_ACTION_MACRO },
+	{ GSKILL_BUTTON_FUNCTION_DPI_UP,             BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_UP) },
+	{ GSKILL_BUTTON_FUNCTION_DPI_DOWN,           BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_DOWN) },
+	{ GSKILL_BUTTON_FUNCTION_CYCLE_DPI_UP,       BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_CYCLE_UP) },
+	{ GSKILL_BUTTON_FUNCTION_CYCLE_PROFILE_UP,   BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_CYCLE_UP) },
+	{ GSKILL_BUTTON_FUNCTION_CYCLE_PROFILE_DOWN, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_DOWN) },
+	{ GSKILL_BUTTON_FUNCTION_DISABLE,            BUTTON_ACTION_NONE },
+};
+
+struct gskill_profile_data {
+	struct gskill_profile_report report;
+	uint8_t res_idx_to_dev_idx[GSKILL_NUM_DPI];
+
+	struct gskill_macro_report macros[GSKILL_BUTTON_MAX];
+};
+
+struct gskill_data {
+	struct gskill_profile_data profile_data[GSKILL_PROFILE_MAX];
+};
+
+static const struct ratbag_button_action *
+gskill_button_function_to_action(enum gskill_button_function_type type)
+{
+	const struct gskill_button_function_mapping *mapping;
+
+	ARRAY_FOR_EACH(gskill_button_function_mapping, mapping) {
+		if (mapping->type == type)
+			return &mapping->action;
+	}
+
+	return NULL;
+}
+
+static enum gskill_button_function_type
+gskill_button_function_from_action(const struct ratbag_button_action *action)
+{
+	const struct gskill_button_function_mapping *mapping;
+
+	ARRAY_FOR_EACH(gskill_button_function_mapping, mapping) {
+		if (ratbag_button_action_match(&mapping->action, action))
+			return mapping->type;
+	}
+
+	return GSKILL_BUTTON_FUNCTION_DISABLE;
+}
+
+static uint8_t
+gskill_calculate_checksum(const uint8_t *buf, size_t len)
+{
+	uint8_t checksum = 0;
+	unsigned i;
+
+	for (i = GSKILL_CHECKSUM_OFFSET + 1; i < len; i++)
+		checksum += buf[i];
+
+	checksum = ~checksum + 1;
+
+	return checksum;
+}
+
+static int
+gskill_general_cmd(struct ratbag_device *device,
+		   uint8_t buf[GSKILL_REPORT_SIZE_CMD]) {
+	int rc;
+	int retries;
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
+				       GSKILL_REPORT_SIZE_CMD,
+				       HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Error while sending command to mouse: %d\n", rc);
+		return rc;
+	}
+
+	for (retries = 0; retries < 10; retries++) {
+		/* Wait for the device to be ready */
+		msleep(20);
```

Comment:
Reiterating here for the github history: The longer wait here is mainly because the mouse seems to be more likely to return a sensible response (e.g. something that actually has the command status filled in) if we wait 20ms as opposed to 10